### PR TITLE
Closed group functionality and synchronisation

### PR DESCRIPTION
**Pull request by RyanRory**
### First time contributor checklist

- [ ] I have read the [README](https://github.com/WhisperSystems/Signal-iOS/blob/master/README.md) and [CONTRIBUTING](https://github.com/WhisperSystems/Signal-iOS/blob/master/CONTRIBUTING.md) documents
- [ ] I have signed the [Contributor Licence Agreement](https://whispersystems.org/cla/)

### Contributor checklist

- [ ] I'm following the [code, UI and style conventions](https://github.com/WhisperSystems/Signal-iOS/blob/master/CONTRIBUTING.md#code-conventions)
- [ ] My commits are rebased on the latest master branch
- [ ] My commits are in nice logical chunks
- [ ] My contribution is fully baked and is ready to be merged as is
- [ ] I have tested my contribution on these devices:
 * iDevice A, iOS X.Y.Z
 * iDevice B, iOS Z.Y

- - - - - - - - - -

### CODE REVIEW

File: Signal/src/Loki/View Controllers/NewClosedGroupVC.swift [hunk 1]
```diff
@@ -169,7 +169,10 @@ final class NewClosedGroupVC : UIViewController, UITableViewDataSource, UITableV
         }
         let userHexEncodedPublicKey = getUserHexEncodedPublicKey()
         let members = [String](selectedContacts) + [ userHexEncodedPublicKey ]
-        let admins = [ userHexEncodedPublicKey ]
+        var admins: [String]!
```
Comment (Mikunj) [inline Signal/src/Loki/View Controllers/NewClosedGroupVC.swift]: Admin should only contain the master device.

File: Signal/src/Loki/View Controllers/NewClosedGroupVC.swift [hunk 2]
```diff
@@ -169,7 +169,10 @@ final class NewClosedGroupVC : UIViewController, UITableViewDataSource, UITableV
         }
         let userHexEncodedPublicKey = getUserHexEncodedPublicKey()
         let members = [String](selectedContacts) + [ userHexEncodedPublicKey ]
```
Comment (Mikunj) [inline Signal/src/Loki/View Controllers/NewClosedGroupVC.swift]: `userHexEncodedPublicKey` is wrong here, we should be adding the master device to the members.

Basically **ALL** closed group logic is centered around the master device, a slave/secondary device public key should never appear in members/admin.

What we should do instead is upon receiving a group message, we should map the sender to their master device and use that to check if they are in the group / admin.

File: SignalServiceKit/src/Messages/OWSMessageSender.m [hunk 3]
```diff
@@ -937,6 +937,23 @@ - (OWSMessageSend *)getSessionRestoreMessageForHexEncodedPublicKey:(NSString *)h
     return [[OWSMessageSend alloc] initWithMessage:message thread:thread recipient:recipient senderCertificate:senderCertificate udAccess:theirUDAccess localNumber:userHexEncodedPublicKey success:^{ } failure:^(NSError *error) { }];
 }
 
+- (LKFriendRequestMessage *)getMultiDeviceFriendRequestMessageForHexEncodedPublicKey:(NSString *)hexEncodedPublicKey inThread:(TSContactThread *)thread transaction:(YapDatabaseReadWriteTransaction *)transaction
```
Comment (gmbnt) [inline SignalServiceKit/src/Messages/OWSMessageSender.m]: This is duplicating a fair bit of code

Comment (RyanRory) [reply SignalServiceKit/src/Messages/OWSMessageSender.m]: Hmm...This should be my dirty solution for the yapdatabase transaction issue. I was facing a transaction in transaction issue...

File: SignalMessaging/contacts/OWSSyncManager.m [hunk 4]
```diff
@@ -307,6 +324,49 @@ - (AnyPromise *)syncContactsForSignalAccounts:(NSArray<SignalAccount *> *)signal
     return promise;
 }
 
+- (AnyPromise *)syncAllGroups
+{
+    NSMutableArray<TSGroupThread *> *groupThreads = @[].mutableCopy;
+    NSMutableArray<AnyPromise *> *promises = @[].mutableCopy;
+    [TSGroupThread enumerateCollectionObjectsUsingBlock:^(id obj, BOOL *stop) {
+        if (![obj isKindOfClass:[TSGroupThread class]]) {
+            if (![obj isKindOfClass:[TSContactThread class]]) {
```
Comment (gmbnt) [inline SignalMessaging/contacts/OWSSyncManager.m]: This check seems like it would be redundant?

Comment (RyanRory) [reply SignalMessaging/contacts/OWSSyncManager.m]: This was copied from the original code of Signal. I tried not to check this and make the `id obj` as `TSGroupThread *obj`, but I got crashes here. So I just left this code as it was.

File: SignalServiceKit/src/Messages/OWSMessageSender.m [hunk 5]
```diff
@@ -937,6 +937,23 @@ - (OWSMessageSend *)getSessionRestoreMessageForHexEncodedPublicKey:(NSString *)h
...
+{
+    thread = [TSContactThread getOrCreateThreadWithContactId:hexEncodedPublicKey transaction:transaction];
```
Comment (gmbnt) [inline SignalServiceKit/src/Messages/OWSMessageSender.m]: This doesn't use the thread that's passed in?

Comment (RyanRory) [reply SignalServiceKit/src/Messages/OWSMessageSender.m]: Oh boy, this line of code is duplicated. This `thread` is exactly the thread that's passed in...

**Pull request merged**